### PR TITLE
node: alter initialization order to avoid panic during statesync

### PR DESCRIPTION
**Pull request by tychoish**
This is a partial fix, as I think statesync can (and does) still fail
to gain enough peers to make progress, but at least it doesn't panic
on startup.

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -378,13 +377,18 @@ func makeNode(
 		eventSinks: eventSinks,
 
 		services: []service.Service{
+			// the order of these services control
```
Comment (creachadair) [inline node/node.go:380]: This comment makes me simultaneously very happy (that it's here) and very sad (that it's necessary). 😆

File: node/node.go [hunk 2]
```diff
@@ -465,12 +469,8 @@ func (n *nodeImpl) OnStart(ctx context.Context) error {
...
 	}
 
-	state, err := n.stateStore.Load()
-	if err != nil {
-		return err
-	}
-	if err := n.evPool.Start(state); err != nil {
-		return err
+	if n.config.Instrumentation.Prometheus && n.config.Instrumentation.PrometheusListenAddr != "" {
+		n.prometheusSrv = n.startPrometheusServer(ctx, n.config.Instrumentation.PrometheusListenAddr)
```
Comment (creachadair) [inline node/node.go:473]: Perhaps:
```suggestion
	if p := n.config.Instrumentation; p.Prometheus && p.PrometheusListenAddr != "" {
		n.prometheusSrv = n.startPrometheusServer(ctx, p.PrometheusListenAddr)
```

File: node/node.go [hunk 3]
```diff
@@ -569,6 +570,23 @@ func (n *nodeImpl) OnStart(ctx context.Context) error {
...
+		if err := reactor.Start(ctx); err != nil {
+			return fmt.Errorf("problem starting service '%T': %w ", reactor, err)
+		}
+	}
+
+	state, err := n.stateStore.Load()
+	if err != nil {
+		return fmt.Errorf("loading state during service start: %w", err)
+	}
+	if err := n.evPool.Start(state); err != nil {
```
Comment (creachadair) [inline node/node.go:586]: Could also probably just `return n.evPool.Start(state)`

**Review by creachadair — APPROVED** [review]

File: internal/consensus/state.go [hunk 4]
```diff
@@ -2036,19 +2037,27 @@ func (cs *State) RecordMetrics(height int64, block *types.Block) {
 func (cs *State) defaultSetProposal(proposal *types.Proposal, recvTime time.Time) error {
 	// Already have one
 	// TODO: possibly catch double proposals
-	if cs.Proposal != nil || proposal == nil {
-		return nil
-	}
+	if shouldReturn, err := func() (bool, error) {
```
Comment (creachadair) [inline internal/consensus/state.go:2040]: It appears `shouldReturn == (err != nil)` in this usage.

Comment (creachadair) [reply internal/consensus/state.go:2040]: Oh, nevermind, I missed a case.

File: internal/consensus/state.go [hunk 5]
```diff
@@ -2060,8 +2069,12 @@ func (cs *State) defaultSetProposal(proposal *types.Proposal, recvTime time.Time
...
 
 	proposal.Signature = p.Signature
-	cs.Proposal = proposal
-	cs.ProposalReceiveTime = recvTime
+	func() {
+		cs.mtx.Lock()
+		defer cs.mtx.RUnlock()
+		cs.Proposal = proposal
+		cs.ProposalReceiveTime = recvTime
+	}()
```
Comment (creachadair) [inline internal/consensus/state.go:2077]: This one would be fine to just bracket, there's no real panic risk.

```suggestion
	cs.mtx.Lock()
	cs.Proposal = proposal
	cs.ProposalReceiveTime = recvTime
	cs.mtx.Unlock()
```

N.B.: Whether or not you like this change, we should fix RUnlock → Unlock.

File: node/node.go (hunk 1 above)
Comment (tychoish) [reply node/node.go:380]: extremely the same

**Comment by tychoish** [comment]
I think that this isn't worth continuing on, as its main goals was to avoid the panic in the e2e tests which we've averted using much less drastic means

**Pull request closed without merging**